### PR TITLE
Allow device adoption

### PR DESCRIPTION
**Pull request by joshuaspence**
Allow for devices to be adopted on creation and forgotten on destruction. As these could be destructive operations I have made them both opt-in with `allow_adopt` and `forget_on_destroy`.

One issue that I haven't fixed is that the acceptance tests fail with `TEST_COUNT=2` (or any number larger than one). The reason for this is that after forgetting a device it seems to disappear completely for about 30 seconds before it reappears. I wanted to think about this problem a bit more and will submit a follow-up PR to fix it.

### CODE REVIEW

File: internal/provider/resource_device.go [hunk 1]
```diff
@@ -85,6 +89,20 @@ func resourceDevice() *schema.Resource {
...
+				Description: "Specifies whether the device should be automatically adopted.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+			},
+			"skip_forget_on_destroy": {
+				Description: "Specifies whether this resource should tell the controller to \"forget\" the device on destroy.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
```
[the remark this answers is not in the document]
Comment (joshuaspence) [reply internal/provider/resource_device.go:104]: Renaming the resource is a good point, but in general that is a destructive operation that requires care. Like if I am managing an EC2 instance or RDS database then without a `state mv` that would involve destroying and recreating my infrastructure.

The end result should be the same either way, it's just that there will be a few seconds (minutes?) where the device is no longer adopted. I'm not sure if the device would retain its settings or not though, hmm. If it does then it's probably fine. If it doesn't then yeah that's not a great idea.

In any case I suspect that deletion is rare and this forget on delete functionality isn't very important. I mainly added it for the acceptance tests. I am fine either way really.

**Comment by joshuaspence** [comment]
Tests should be fixed by #190

**Comment by joshuaspence** [comment]
Needs a bit more work. It looks harmless, but I see some warnings/errors in the controller logs:

```
[2021-09-26T02:09:48,619] <webapi-243> WARN  sanitize - Invalid key exists in Device payload, key=mac
[2021-09-26T02:09:48,619] <webapi-243> WARN  sanitize - Invalid key exists in Device payload, key=adopted
[2021-09-26T02:09:48,620] <webapi-243> WARN  sanitize - Invalid key exists in Device payload, key=state
[2021-09-26T02:09:55,309] <fake-UDC48X6> ERROR dev    - dev[00:27:22:00:00:06] failed to update capability: api.err.InvalidReportPort
```

**Comment by paultyng** [comment]
Those sanitize logs are interesting though, I wonder if we could incorporate that output somehow in to the acc tests themselves.

**Pull request merged**